pupil_src/shared_modules/video_capture/file_backend.py

```diff
@@ -307,43 +387,35 @@ def get_frame(self):
         except IndexError:
             logger.info("Reached end of timestamps list.")
             raise EndofVideoError("Reached end of timestamps list.")
-
         self.target_frame_idx = index + 1
         self.current_frame_idx = index
-
         return Frame(timestamp, frame, index=index)
 
     @ensure_initialisation(fallback_func=lambda evt: sleep(0.05))
     def recent_events_external_timing(self, events):
         try:
             last_index = self._recent_frame.index
         except AttributeError:
-            # called once on start when self._recent_frame is None
+            # Get frame at beginnning
+            frame = self.get_next_frame()
+            self._recent_frame = frame
             last_index = -1
 
-        frame = None
+        # Seek Frame
         pbt = self.g_pool.seek_control.current_playback_time
         ts_idx = self.g_pool.seek_control.ts_idx_from_playback_time(pbt)
-        if ts_idx == last_index:
-            frame = self._recent_frame.copy()
-            if self.play and ts_idx == self.get_frame_count() - 1:
-                logger.info("Video has ended.")
-                self.g_pool.seek_control.play = False
-
-        elif ts_idx < last_index or ts_idx > last_index + 1:
-            # time to seek
+        if ts_idx < last_index or ts_idx > last_index + 1:
             self.seek_to_frame(ts_idx)
-
-        # Only call get_frame() if the next frame is actually needed
-        from time import monotonic
-
-        try:
-            frame = frame or self.get_frame()
-        except EndofVideoError:
-            logger.info("Video has ended.")
-            self.g_pool.seek_control.play = False
-            frame = frame or self._recent_frame.copy()
-
+        # Normla Case to get next frame
+        if not self.play:
+            frame = self._recent_frame
```

Comment:
We might need to make a copy here since there are plugins that modify the decoded image buffer. I will try to confirm this in my tests later.

Reply:
Which one we should make a copy here?

Reply:
`frame = self._recent_frame.copy()`